### PR TITLE
Improve generic code in `type_coverage.h`

### PR DESCRIPTION
**Pull request by andreyromanof**
This PR provides a new function `for_all_combinations` that can combinate types, values, etc.
Refactor old functions like `for_all_type` for removing unused variable and dummy values in fold expression.
Also made constructor of `named_type_pack` private. Instead of a constructor, there is the factory function `generate_named`. This change will allow us to freely change attributes and the way to construct the actual type pack inside the class without changing existing tests.

The function `for_all_combinations` will be used in new accessor tests.

Depends on:
- #308

### CODE REVIEW

File: tests/common/type_coverage.h [hunk 1]
```diff
@@ -2,7 +2,7 @@
 //
 //  SYCL 2020 Conformance Test Suite
 //
-// Provide common functions for type coverage
+//  Provide common functions for type coverage
```
Comment (bader) [inline tests/common/type_coverage.h:5]: Minor: Is it a change made by clang-format? If so, I think it tries to align with the line 3, so I would remove one space there before SYCL instead.

File: tests/common/type_coverage.h [hunk 2]
```diff
@@ -14,14 +14,14 @@
...
+
 /**
  * @brief Retrieve type name; by default just forward the given one
  */
 template <typename T>
 struct type_name_string {
-    static std::string get(std::string dataType) {
-        return dataType;
-    }
+  static std::string get(std::string dataType) { return dataType; }
```
Comment (bader) [inline tests/common/type_coverage.h:24]: @andreyromanof, please, refrain from adding code formatting changes for the parts, which logic you don't modify. It makes code review difficult. Please, move such changes to a separate PR. It would much easier to review this patch if there is a clear separation between code formatting (non-functional) changes and changes "improving" `type_coverage.h`.

Comment (andreyromanof) [reply tests/common/type_coverage.h:24]: I moved format changes to separate PR (#308), to simplify the review process for this PR.

Comment (bader) [reply tests/common/type_coverage.h:24]: Do you mean we should hold reviews until #308 is merged?

Comment (andreyromanof) [reply tests/common/type_coverage.h:24]: I think we should because of the large changes in format. 
I converted PR to draft until #308 is merged.

**Comment by andreyromanof** [comment]
I can't edit the reviewers so I duplicate:
>I think we should hold reviews because of the large changes in format.
>I converted PR to draft until https://github.com/KhronosGroup/SYCL-CTS/pull/308 is merged.

**Review by ProGTX — APPROVED** [review]
